### ISSUE TITLE
Coq Anomaly  using "with" construct

### ISSUE THREAD

**Issue by TimRichter** [mattam82/Coq-Equations#70]
Hi,

at the very end the code below gives this Coq warning:

    Induction principle could not be proved automatically: 
    Anomaly
    "File "src/principles.ml", line 433, characters 16-22: Assertion failed." 
    Please report at http://coq.inria.fr/bugs/.

I quite often encounter this warning when trying to use "with". Do I do
something wrong? Should I rather report it in the Coq list (as it says)?

Best,
Tim

```coq
Require Import Fin.
From Equations Require Import Equations.
Set Equations Transparent.

Equations FL (n : nat) : Fin.t (S n) :=
  FL 0     := F1;
  FL (S n) := FS (FL n).

Equations FU {n : nat} (x : Fin.t n) : Fin.t (S n) :=
  FU F1     := F1;
  FU (FS x) := FS (FU x).

Equations invertFin {n : nat} (x : Fin.t n) : Fin.t n :=
  invertFin  F1    := FL _;
  invertFin (FS x) := FU (invertFin x).

Equations invFULemma {n : nat} (x : Fin.t n) :
                     invertFin (FU x) = FS (invertFin x) :=
  invFULemma F1     := _;
  invFULemma (FS x) := (f_equal _ (invFULemma x)).

Equations invFLLemma (n : nat) : invertFin (FL n) = F1 :=
  invFLLemma 0     := eq_refl;
  invFLLemma (S n) := (f_equal _ (invFLLemma n)).

Equations invertFinInv {n : nat} (x : Fin.t n) :
                       invertFin (invertFin x) = x :=
  invertFinInv {n:=0}      x     :=! x;
  invertFinInv {n:=(S _)}  F1    := (invFLLemma _);
  invertFinInv {n:=(S _)} (FS y) := (eq_trans
                                     (invFULemma (invertFin y))
                                     (f_equal _ (invertFinInv y))).

Definition invFinViewType {n : nat} (x : (Fin.t n)) : Type :=
  { y : Fin.t n & x = invertFin y }.

Definition invFinView {n : nat} (x : (Fin.t n)) : invFinViewType x :=
  existT _ (invertFin x) (eq_sym (invertFinInv x)).

Equations finFUOrFL {n : nat} (x : Fin.t (S n)) :
                    { y : Fin.t n & x = FU y } + ( x = FL n ) :=
  finFUOrFL  {n:=0}     F1                     := (inr eq_refl);
  finFUOrFL  {n:=(S _)} x <= invFinView x => {
                          | (existT F1 eq)     := (inr eq);
                          | (existT (FS _) eq) := (inl (existT _ (invertFin _) eq))}.
```

**Comment by mattam82** [comment]
That's quite strange because  finFUOrFL_ind_fun is actually defined (at least in the 8.7 version I'm using). It's probably the generation of the elimination principle from that induction principle that is buggy here, I'll investigate.

**Comment by mattam82** [comment]
Also one « workaround » is to use the (noind) flag which will instruct
Equations not to generate the graph and elimination principle, if this is
ever blocking.
Le jeu. 12 avr. 2018 à 21:18, Tim Richter <notifications@github.com> a
écrit :

**Comment by TimRichter** [comment]
Yes, finFUOrFL_ind_fun is also defined in 8.7.2. 

The warning as such is not a problem. I get the same though anytime I use finFUOrFL in a with ...

And as I don't fully understand the rewriting magic behind equations (which I enjoy!), I'm afraid 
that the fact something could not be generated could weaken the reasoning powers?

For now, I'll just ignore the warnings then.

Thank you,
Tim

**Comment by mattam82** [comment]
I found the bug, fixed by https://github.com/mattam82/Coq-Equations/tree/fix70

**Comment by mattam82** [comment]
The fix is in the 8.8 branch as well. I'll make a bugfix release soon.
